### ISSUE TITLE
Unable to create assymetric key encryption on Config Server

### ISSUE THREAD

**Issue by da3mon-01** [spring-cloud/spring-cloud-config#539]
Hi!

I am working on creating a configserver that can use the key management solution with an RSA keypair detailed [here](http://cloud.spring.io/spring-cloud-static/Brixton.SR6/#_key_management), but I cannot seem to get it to work.
My expected workflow: I add the public key to the java keystore for the application to the config server, and use the /encrypt endpoint to add encrypt the property. Then I mount the private key in the client application which will decrypt it locally.

Here is what I did:
Create a Config Server with the following properties:

``` server.port=8888
spring.cloud.config.server.git.uri=${HOME}/repos/config-repo
encrypt.keyStore.location=classpath:/server.jks
encrypt.keyStore.alias=encrypt
encrypt.keyStore.password=letmein
encrypt.keyStore.secret=changepass
spring.cloud.config.server.encrypt.enabled=false
```

Created the keypair, then exported the public key as a certificate into server.jks, which is created with the import command.

```
[da3mon@localhost resources]$ keytool -genkeypair -alias encrypt -keyalg RSA -validity 360 -keystore encrypt3.jks -storepass letmein -keypass changepass
What is your first and last name?
  [Unknown]:  Config
What is the name of your organizational unit?
  [Unknown]:  Server
What is the name of your organization?
  [Unknown]:  App 
What is the name of your City or Locality?
  [Unknown]:  Budapest
What is the name of your State or Province?
  [Unknown]:  Budapest
What is the two-letter country code for this unit?
  [Unknown]:  HU
Is CN=Config, OU=Server, O=App, L=Budapest, ST=Budapest, C=HU correct?
  [no]:  yes

[da3mon@localhost resources]$ keytool -export -alias encrypt -rfc -keystore encrypt3.jks -storepass letmein -keypass changepass -file encrypt3.cer
Certificate stored in file <encrypt3.cer>

[da3mon@localhost resources]$ keytool -import -file encrypt3.cer -alias encrypt -keystore server.jks -storepass letmein -keypass changepass
Owner: CN=Config, OU=Server, O=App, L=Budapest, ST=Budapest, C=HU
Issuer: CN=Config, OU=Server, O=App, L=Budapest, ST=Budapest, C=HU
Serial number: 4f1d48ec
Valid from: Tue Oct 25 11:18:28 CEST 2016 until: Fri Oct 20 11:18:28 CEST 2017
Certificate fingerprints:
     MD5:  DA:CD:F7:59:41:67:92:72:BB:FC:09:D7:CB:79:52:78
     SHA1: 40:93:B3:51:19:2E:DF:CA:86:DF:5B:31:E4:E8:1F:7D:4B:D7:B2:68
     SHA256: 86:01:46:ED:4F:96:76:0C:4F:2B:22:AA:6E:9F:97:27:5D:D3:81:E4:12:CC:60:66:0D:05:A9:96:C7:C7:2C:12
     Signature algorithm name: SHA256withRSA
     Version: 3

Extensions: 

#1: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: 74 43 71 2D 02 A0 4C 9D   96 19 61 9C EF 45 DC 66  tCq-..L...a..E.f
0010: B5 80 53 76                                        ..Sv
]
]

Trust this certificate? [no]:  yes
Certificate was added to keystore
[da3mon@localhost resources]$ keytool -list -keystore server.jks
Enter keystore password:  

Keystore type: JKS
Keystore provider: SUN

Your keystore contains 1 entry

encrypt, 2016.10.25., trustedCertEntry, 
Certificate fingerprint (SHA1): 40:93:B3:51:19:2E:DF:CA:86:DF:5B:31:E4:E8:1F:7D:4B:D7:B2:68
```

However when I check whether I can encrypt variables, I am unable to do so:

```
[da3mon@localhost resources]$ curl localhost:8888/encrypt/status
{"timestamp":1477387291480,"status":500,"error":"Internal Server Error","exception":"java.lang.IllegalStateException","message":"Cannot load keys from store: class path resource [server.jks]","path":"/encrypt/status"}[da3mon@localhost resources]$`
```

Any Ideas what I am doing wrong?

EDIT: 
I did some research:
According to the manual:

The encryption is done with the public key, and a private key is needed for decryption. Thus in principle you can configure only the public key in the server if you only want to do encryption (and are prepared to decrypt the values yourself locally with the private key). In practice you might not want to do that because it spreads the key management process around all the clients, instead of concentrating it in the server. On the other hand it's a useful option if your config server really is relatively insecure and only a handful of clients need the encrypted properties.

However KeyStoreKeyFactory seems to require the private key anyway:
https://github.com/dsyer/spring-security-rsa/blob/master/src/main/java/org/springframework/security/rsa/crypto/KeyStoreKeyFactory.java

```
public KeyPair getKeyPair(String alias, char[] password) {
        try {
            synchronized (lock) {
                if (store == null) {
                    synchronized (lock) {
                        store = KeyStore.getInstance("jks");
                        store.load(resource.getInputStream(), this.password);
                    }
                }
            }
            RSAPrivateCrtKey key = (RSAPrivateCrtKey) store.getKey(alias, password);
            RSAPublicKeySpec spec = new RSAPublicKeySpec(key.getModulus(),
                    key.getPublicExponent());
            PublicKey publicKey = KeyFactory.getInstance("RSA").generatePublic(spec);
            return new KeyPair(publicKey, key);
        }
        catch (Exception e) {
            throw new IllegalStateException("Cannot load keys from store: " + resource, e);
        }
    }
```

Is this a doc issue then?

**Comment by ryanjbaxter** [comment]
You can keep all encryption/decryption on the server, in fact that is the default behavior.

**Comment by da3mon-01** [comment]
Hi Ryan!

Sorry for the late response, but I did manage to figure out a solution, and just forgot about this ticket :/ I needed to implement 2 config servers: One has full encryption functionality, and is used to encrypt passwords, before storing them in git. The second serves the actual configuration, and does not have any encrypt/decrypt functionality. The actual decryption is handled by the spring-config-client in the application itself.
This was a requirement from ITSec so that different banking apps of ours use different keys, thanks for your answer anyway!
